Dockerfile

```diff
@@ -1,26 +1,50 @@
+ARG HOME=/app UID=9500 GID=9500 VERSION="base"
+ARG NODE_ENV=production
+
+# Define the base image, here we are root.
 FROM node:18 as base
 
-WORKDIR /app
+RUN touch /is-docker
+
+# Define the user and working directory, everything from this is user land
+FROM base as user
+
+ARG HOME UID GID
+
+RUN <<EOF
+groupadd -g ${GID} user
+useradd -u ${UID} -g ${GID} -s /sbin/nologin -d ${HOME} user
+EOF
+
+USER user
+WORKDIR ${HOME}
+
+FROM user as build
 
-RUN --mount=type=bind,src=package.json,dst=/app/package.json \
-  npm install
+ARG HOME
+# Always development for building as we need to build the code
+ENV NODE_ENV=development
 
 RUN \
-  --mount=type=bind,src=package.json,dst=/app/package.json \
-  --mount=type=bind,src=src,dst=/app/src/ \
-  --mount=type=bind,src=tsconfig.json,dst=/app/tsconfig.json \
-  npm run build
+  --mount=type=bind,source=src,target=${HOME}/src \
+  --mount=type=bind,source=package.json,target=${HOME}/package.json \
+  --mount=type=bind,source=package-lock.json,target=${HOME}/package-lock.json \
+  --mount=type=bind,source=tsconfig.json,target=${HOME}/tsconfig.json \
+  <<EOF
+npm install --loglevel=verbose --no-save
+npm run build
+EOF
 
-FROM base as final
+FROM user as final
 
-ARG VERSION="base"
+ARG HOME VERSION NODE_ENV
 
-ENV VERSION=${VERSION}
+ENV VERSION=${VERSION} NODE_ENV=${NODE_ENV}
```

Comment:
npm install will respect node_env allowing us to build the container either with prod dependencies only, or with dev also, just by specifying a build arg. kind of neat.